source/common/router/router.cc

```diff
@@ -1976,6 +2011,9 @@ void Filter::doRetry(bool can_send_early_data, bool can_use_http3, TimeoutRetry
   }
   UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(
       *this, std::move(generic_conn_pool), can_send_early_data, can_use_http3);
+  if (retry_state_ && retry_state_->shouldRetryOnConnectionFailure() && wait_for_connect_) {
+    upstream_request->addUpstreamCallbacks(*this);
```

Comment:
do we still need to add to the upstream callbacks now?

Reply:
Yes, otherwise how can we know that we got a connection with the upstream.
cf linge 862 above `onUpstreamConnectionEstablished`: https://github.com/DataDog/envoy/blob/c0e00f3a9f0d21b76a6d14c98efbc65341ed9790/source/common/router/router.cc#L862